### ISSUE TITLE
package navidrome WebUI as Progressive Web Application

### ISSUE THREAD

**Issue by certuna** [deluan/navidrome#498]
I'm a healthy sceptic of web apps as the eternal Next Big Thing™, however I do think the navidrome WebUI is awesome, and should be packageable as a PWA with relatively little effort.

Advantages:
- allows your navidrome server to run as a 'native-looking' app on Android, Windows 10, macOS, ChromeOS and (with limitations) iOS.
- allows navidrome to implement new features (multi-value tags, more metadata fields) without resorting to breaking/extending the SubSonic API
- allows for offline caching, which could be used for sync-to-device or pre-loading

Disadvantages:
- https required, which at the moment would limit it to navidrome running behind a tls-capable proxy
- PWA's on iOS are crippled and the web is [filled](https://www.freecodecamp.org/news/i-built-a-pwa-and-published-it-in-3-app-stores-heres-what-i-learned-7cb3f56daf9b/) with [developer rants](https://blog.prototyp.digital/what-we-learned-about-pwas-and-audio-playback/) how it is terrible. This is true, however iOS 14+ might see some more improvements, and PWA's on iOS being suboptimal doesn't invalidate the benefits for Android and Windows.

**Comment by deluan** [comment]
I thought about this before but didn't get too far because of the poor PWA support in iOS. If Apple adds proper support for PWA's in iOS, then I feel this is something worth doing. If not, I rather work (or contribute) on a cross-platform client, but then it would be only after Navidrome reaches 1.0

**Comment by certuna** [comment]
Writing fully functional native clients for Windows, Android and iOS is probably a hell of a lot more work than writing a PWA manifest and some service worker code, so this might be relatively "low hanging fruit"?

The situation with iOS is regrettable, and I can imagine your hesitation. But even without proper support, Android + Windows 10 + macOS is still a target audience big enough to consider I think.

**Comment by certuna** [comment]
Bummer: [background audio still not working](https://dev.to/wes566/ios-14-pwa-improvements-1o3e) in iOS 14. At least it was [acknowledged ](https://bugs.webkit.org/show_bug.cgi?id=198277#c10)by Apple.

**Comment by deluan** [comment]
Hey, do you think #518 was enough to close this? I mean, implementing ex: offline cache and making the UI an app on its own would be basically a new project.

**Comment by certuna** [comment]
I think so yes, this can be closed - with the correct manifest, the "publishing" of Navidrome as a PWA is done. I might put in an issue to implement offline caching of metadata and media in IndexedDB (this would basically bring full-on sync to the WebUI), but that's indeed another thing.

By the way, I've just done some more testing with pinnin Navidrome as a PWA on iOS 14, now that I've set up a reverse proxy with https. Contrary to earlier tests, background audio does seem to work, I can queue up a full album, and the app happily moves on to the next track even with the app minimized and/or the screen turned off. For some reason the lockscreen playback controls display `[object Object]` as track name (and a Safari icon) but otherwise the experience is very good.

**Comment by deluan** [comment]
> For some reason the lock screen playback controls display [object Object] instead of the track name

I think this is caused by the way I set the track name for the media player... Can you open a new issue, with a screenshot, so I don't forget about this? Thanks a bunch!
